File: backend_python/src/models/media.py
"""
MediaAsset Models (Discriminated Union Pattern).

This module defines strictly typed media asset models for all upload types.
Uses Pydantic V2 Discriminated Unions via `Literal` type discriminators.

**Golden Sync**: TypeScript interfaces must be generated in `web_client/types/media.ts`.
"""
from typing import Literal, Optional, Union
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
import uuid


class MediaAssetBase(BaseModel):
    """
    Base model for all media assets.
    Common fields shared across image, video, and document types.
    """
    model_config = ConfigDict(extra="forbid")  # Strict: No extra fields

    id: str = Field(default_factory=lambda: uuid.uuid4().hex, description="Unique asset ID")
    url: str = Field(..., description="Public URL to access the asset")
    filename: str = Field(..., description="Original sanitized filename")
    mime_type: str = Field(..., description="Validated MIME type from Magic Bytes")
    size_bytes: int = Field(..., ge=0, description="File size in bytes")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="Upload timestamp")


class ImageMediaAsset(MediaAssetBase):
    """
    Image-specific asset model.
    
    Discriminator: asset_type = "image"
    """
    asset_type: Literal["image"] = "image"
    
    # Image-specific metadata
    width: Optional[int] = Field(None, ge=1, description="Image width in pixels")
    height: Optional[int] = Field(None, ge=1, description="Image height in pixels")
    
    # Storage paths
    file_path: str = Field(..., description="Storage path in bucket")
    signed_url: Optional[str] = Field(None, description="Signed URL (expires)")


class VideoMediaAsset(MediaAssetBase):
    """
    Video-specific asset model.
    
    Discriminator: asset_type = "video"
    """
    asset_type: Literal["video"] = "video"
    
    # Video-specific metadata
    width: Optional[int] = Field(None, ge=1, description="Video width in pixels")
    height: Optional[int] = Field(None, ge=1, description="Video height in pixels")
    duration_seconds: Optional[float] = Field(None, ge=0, description="Duration in seconds")
    
    # File API reference (for Gemini multimodal processing)
    file_uri: str = Field(..., description="Google AI File API URI")
    state: str = Field(default="ACTIVE", description="File API processing state")


class DocumentMediaAsset(MediaAssetBase):
    """
    Document-specific asset model (PDF, etc).
    
    Discriminator: asset_type = "document"
    """
    asset_type: Literal["document"] = "document"
    
    page_count: Optional[int] = Field(None, ge=1, description="Number of pages (PDF)")
    
    # Storage paths
    file_path: str = Field(..., description="Storage path in bucket")
    signed_url: Optional[str] = Field(None, description="Signed URL (expires)")


# ============================================================================
# DISCRIMINATED UNION TYPE
# ============================================================================

MediaAsset = Union[ImageMediaAsset, VideoMediaAsset, DocumentMediaAsset]
"""
Discriminated Union type for all media assets.

Usage in FastAPI responses:
    response_model=ImageMediaAsset  # For specific endpoint
    response_model=MediaAsset       # For polymorphic responses

The discriminator field is `asset_type`:
- "image" -> ImageMediaAsset
- "video" -> VideoMediaAsset  
- "document" -> DocumentMediaAsset
"""


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def parse_media_asset(data: dict) -> MediaAsset:
    """
    Parse a dictionary into the correct MediaAsset subtype.
    
    Uses the `asset_type` discriminator to determine the model.
    
    Args:
        data: Dictionary with asset data including `asset_type`
        
    Returns:
        Appropriate MediaAsset subtype instance
        
    Raises:
        ValueError: If asset_type is missing or invalid
    """
    asset_type = data.get("asset_type")
    
    if asset_type == "image":
        return ImageMediaAsset.model_validate(data)
    elif asset_type == "video":
        return VideoMediaAsset.model_validate(data)
    elif asset_type == "document":
        return DocumentMediaAsset.model_validate(data)
    else:
        raise ValueError(f"Unknown asset_type: {asset_type}")

File: web_client/types/media.ts
/**
 * MediaAsset Types (Golden Sync with backend_python/src/models/media.py)
 *
 * This file contains TypeScript interfaces that mirror the Python Pydantic models.
 * ANY CHANGES TO THE BACKEND MODELS MUST BE REFLECTED HERE.
 *
 * @see backend_python/src/models/media.py
 */

/**
 * Base fields common to all media asset types.
 */
interface MediaAssetBase {
    /** Unique asset ID (UUID hex) */
    id: string;
    /** Public URL to access the asset */
    url: string;
    /** Original sanitized filename */
    filename: string;
    /** Validated MIME type from Magic Bytes */
    mime_type: string;
    /** File size in bytes */
    size_bytes: number;
    /** ISO timestamp of upload */
    created_at?: string;
}

/**
 * Image-specific asset.
 * Discriminator: asset_type = "image"
 */
export interface ImageMediaAsset extends MediaAssetBase {
    asset_type: 'image';
    /** Image width in pixels */
    width?: number;
    /** Image height in pixels */
    height?: number;
    /** Storage path in bucket */
    file_path: string;
    /** Signed URL (expires) */
    signed_url?: string;
}

/**
 * Video-specific asset.
 * Discriminator: asset_type = "video"
 */
export interface VideoMediaAsset extends MediaAssetBase {
    asset_type: 'video';
    /** Video width in pixels */
    width?: number;
    /** Video height in pixels */
    height?: number;
    /** Duration in seconds */
    duration_seconds?: number;
    /** Google AI File API URI */
    file_uri: string;
    /** File API processing state */
    state: string;
}

/**
 * Document-specific asset (PDF, etc).
 * Discriminator: asset_type = "document"
 */
export interface DocumentMediaAsset extends MediaAssetBase {
    asset_type: 'document';
    /** Number of pages (PDF) */
    page_count?: number;
    /** Storage path in bucket */
    file_path: string;
    /** Signed URL (expires) */
    signed_url?: string;
}

/**
 * Discriminated union type for all media assets.
 *
 * Use `asset_type` field to narrow the type:
 * ```ts
 * if (asset.asset_type === 'image') {
 *   console.log(asset.file_path); // OK
 * }
 * ```
 */
export type MediaAsset = ImageMediaAsset | VideoMediaAsset | DocumentMediaAsset;

/**
 * Type guard for ImageMediaAsset
 */
export function isImageAsset(asset: MediaAsset): asset is ImageMediaAsset {
    return asset.asset_type === 'image';
}

/**
 * Type guard for VideoMediaAsset
 */
export function isVideoAsset(asset: MediaAsset): asset is VideoMediaAsset {
    return asset.asset_type === 'video';
}

/**
 * Type guard for DocumentMediaAsset
 */
export function isDocumentAsset(asset: MediaAsset): asset is DocumentMediaAsset {
    return asset.asset_type === 'document';
}

// =============================================================================
// UPLOAD ITEM STATE (Frontend-only, used by useUpload hook)
// =============================================================================

export type UploadStatus = 'idle' | 'compressing' | 'uploading' | 'success' | 'error';

/**
 * Frontend upload state item.
 * Tracks the lifecycle of a file from selection to completion.
 */
export interface UploadItem {
    /** Unique client-side ID (UUID) */
    id: string;
    /** Original File object */
    file: File;
    /** Local blob URL for preview */
    previewUrl: string;
    /** Upload progress (0-100) */
    progress: number;
    /** Current status */
    status: UploadStatus;
    /** Error message if status === 'error' */
    error?: string;
    /** Server response data when upload completes */
    serverData?: MediaAsset;
    /** AbortController for cancellation */
    abortController?: AbortController;
}
